### PR TITLE
[FLINK-24580][Connectors/Kinesis] Make ConnectTimeoutException recoverable

### PR DESCRIPTION
**Pull request by john-karp**
## What is the purpose of the change

High-volume Kinesis consumers can sporadically get Connect errors when connecting to the Kinesis endpoint.
Current behavior is to consider this a fatal error, causing Flink to roll back to previous checkpoint. However retries will typically lead to success, so this patch causes connect errors to be considered recoverable.

## Brief change log

- Make ConnectTimeoutException recoverable
- Unit test for above

## Verifying this change

- New unit test
- Applying this patch has allowed one of our Flink jobs to run continuously, instead of failing several times per day.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by dannycranmer** [comment]
@john-karp can you please rebase to allow CI to run?

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java [hunk 1]
```diff
@@ -420,6 +421,8 @@ private String getShardIterator(GetShardIteratorRequest getShardIteratorRequest)
     protected boolean isRecoverableSdkClientException(SdkClientException ex) {
         if (ex instanceof AmazonServiceException) {
             return KinesisProxy.isRecoverableException((AmazonServiceException) ex);
+        } else if (ex.getCause() instanceof ConnectTimeoutException) {
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:424]: nit: You could use `ExceptionUtils#hasCause` incase the exception gets wrapped:
- https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/exception/ExceptionUtils.html#hasCause-java.lang.Throwable-java.lang.Class-

Comment (john-karp) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:424]: How about having the test be`ExceptionUtils.findThrowable(ex, ConnectException.class).isPresent()`? Using `java.net.ConnectException` has the advantage that it is independent of the HTTP library used by the AWS SDK. (The `ExceptionUtils` here is from `org.apache.flink.util`).

Or, using `ExceptionUtils` from `org.apache.commons.lang3`, the test could be `ExceptionUtils.getRootCause(ex) instanceof ConnectException`.

Comment (dannycranmer) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java:424]: Yes I agree, either approach is fine. I will wait for an update and merge once tests pass

Comment (john-karp) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxy.java]: Updated, tests pass.

**Pull request merged**